addons/utm/models/utm.py

```diff
@@ -42,14 +49,88 @@ def _group_expand_stage_ids(self, stages, domain, order):
         stage_ids = stages._search([], order=order, access_rights_uid=SUPERUSER_ID)
         return stages.browse(stage_ids)
 
+    def _merge_utm_campaigns(self, campaign=False, name=False, user_id=False, stage_id=False, tag_ids=False):
+        if len(self.ids) <= 1:
+            raise UserError(_('Please select more than one campaign from the list.'))
+        if not all(r.active for r in self):
+            raise UserError(_('Only active campaigns can be merged.'))
+        if campaign and campaign not in self:
+            raise UserError(_('Kept campaign should be one of the merged campaign.'))
+
+        merged_campaign = campaign or self[0]
+
+        merged_data = self._merge_data()
```

Comment:
I would not extract to a merge_data method, it adds complexity.

Suggestion:

Instead of passing a bunch of parameters to your method, rewrite it as:
`def _merge_utm_campaigns(self, merge_into_campaign, merge_values={})`

Then you can:
```python
default_merge_values = {
    'user_id': self[0].user_id
    'stage_id': self.env['utm.stage'].search([], limit=1) # do you need to do this? the default will take care of it?
    'tag_ids': reduce(lambda tag_a, tag_b: tag_a | tag_b, (campaign.tag_ids for campaign in self))
    'is_website': any(campaign.is_website for campaign in self)
}

default_merge_values.update(merge_values)
merge_into_campaign.write(default_merge_values)
```

Should be easier to read.

Reply:
update()! That's cool. Much more readable.

Reply:
Seems update is not suitable for our case. The merge_values may contain None values that is not allowed, I can not simply update them, I need to check the data is not None.